---
#title: "Exome assembly stats"
#author: "gwct"
#date: "08/28/2020"
output: 
  html_document:
    includes:
      in_header: '../html-chunks/rmd_header.html'
    df_print: paged
    code_folding: hide
---

```{r setup, warning=FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)

library(ggplot2)
library(ggbeeswarm)
library(ggpubr)
source("../lib/design.r")

targets = 203198
target_avg_len = 267.5448577249776
target_min_len = 100
target_total_len = 54364580

tiles = 470359
tile_avg_len = 109.8747
tile_min_len = 57
tile_total_len = 51679441
# Target and tile info

htmltools::includeHTML("../html-chunks/nav.html")
```

# Exome assembly stats with Spades
### Number of scaffolds by batch

```{r fig1, out.width="100%", fig.align="center"}
spades_data = read.csv("../../../data/exome-stats.csv", header=TRUE)
#spades_data = subset(spades_data, "Mus pos_ctrl" !%in% Species)
#spades_data$batch = as.character(spades_data$batch)
spades_data$Batch = factor(spades_data$Batch, levels=c("1", "2", "3", "4", "5", "6"))

####################
# Spades contig count distributions by batch
scaff_batch_p = ggplot(spades_data, aes(x=Batch, y=count, fill=Batch)) +
  geom_hline(yintercept=targets, linetype=2, size=1, color="#666666") +
  geom_hline(yintercept=tiles, linetype=2, size=1, color="#9494b8") +
  geom_violin(width=1, color="#666666", size=0.3) +
  geom_quasirandom(size=0.75, alpha=0.6, width=0.15, color="#999999") +
  scale_fill_manual(labels=c("Original 40", "Rattus 8", "Mus 14", "New 70", "Aussie 46", "New 27"),
                     values=c("#920000", "#333333", "#006ddb", "#db6d00", "#b66dff", "#db6d00")) +
  annotate("text", x=1.1, y=targets-10000, label="# targets", color="#666666", size=4) +
  annotate("text", x=0.9, y=tiles-10000, label="# tiles", color="#9494b8", size=4) +
  #ggtitle("Spades assembly counts") +
  labs(x="Sequencing batch", y="# Scaffolds") +
  scale_x_discrete(labels=c("Original 40", "Rattus 8", "Mus 14", "New 70", "Aussie 46", "New 27")) +
  bartheme() + 
  theme(legend.position="none",
        axis.text.x = element_text(angle=25, hjust=1))

print(scaff_batch_p)
```

### Scaffold length by batch

```{r fig2, out.width="100%", fig.align = "center"}
####################
# Spades length distributions, by batch
len_batch_p2 = ggplot(spades_data, aes(x=Batch, y=avg.len, fill=Batch)) +
  geom_hline(yintercept=target_avg_len, linetype=2, size=1, color="#666666") +
  geom_hline(yintercept=target_min_len, linetype=2, size=1, color="#666666") +
  geom_hline(yintercept=tile_avg_len, linetype=2, size=1, color="#9494b8") +
  geom_hline(yintercept=tile_min_len, linetype=2, size=1, color="#9494b8") +
  geom_violin(width=1, color="#666666", size=0.3) +
  geom_quasirandom(size=1, alpha=0.6, width=0.15, color="#999999") +
  scale_fill_manual(labels=c("Original 40", "Rattus 8", "Mus 14", "New 70", "Aussie 46", "New 27"),
                     values=c("#920000", "#333333", "#006ddb", "#db6d00", "#b66dff", "#db6d00")) +
  annotate("text", x=1.4, y=target_avg_len-25, label="Avg. target len", color="#666666", size=4) +
  annotate("text", x=1.4, y=target_min_len-20, label="Min. target len", color="#666666", size=4) +
  annotate("text", x=5.5, y=tile_avg_len+35, label="Avg. tile len", color="#9494b8", size=4) +
  annotate("text", x=5.5, y=tile_min_len-15, label="Min. tile len", color="#9494b8", size=4) +
  #ggtitle("Spades assembly counts") +
  labs(x="Sequencing batch", y="Avg. scaffold length") +
  scale_x_discrete(labels=c("Original 40", "Rattus 8", "Mus 14", "New 70", "Aussie 46", "New 27")) +
  bartheme() +
  theme(legend.position="none",
        axis.text.x = element_text(angle=25, hjust=1))

print(len_batch_p2)
```

### Read depth to assembly by batch

```{r fig3, out.width="100%", fig.align = "center", warning=FALSE}
####################
# Spades length distributions, by batch
depth_p = ggplot(spades_data, aes(x=Batch, y=avg.depth.per.base..samtools., fill=Batch)) +
  geom_violin(width=1, color="#666666", size=0.3) +
  geom_quasirandom(size=1, alpha=0.6, width=0.15, color="#999999") +
  scale_fill_manual(labels=c("Original 40", "Rattus 8", "Mus 14", "New 70", "Aussie 46", "New 27"),
                     values=c("#920000", "#333333", "#006ddb", "#db6d00", "#b66dff", "#db6d00")) +
  #ggtitle("Spades assembly counts") +
  labs(x="Sequencing batch", y="Avg. depth to assembly") +
  scale_x_discrete(labels=c("Original 40", "Rattus 8", "Mus 14", "New 70", "Aussie 46", "New 27")) +
  bartheme() +
  theme(legend.position="none",
        axis.text.x = element_text(angle=25, hjust=1))

print(depth_p)
```

### Read depth to mouse by batch

```{r fig4, out.width="100%", fig.align = "center", warning=FALSE}
####################
# Spades length distributions, by batch
depth_p = ggplot(spades_data, aes(x=Batch, y=avg.depth.per.base.to.mouse.targets..samtools., fill=Batch)) +
  geom_violin(width=1, color="#666666", size=0.3) +
  geom_quasirandom(size=1, alpha=0.6, width=0.15, color="#999999") +
  scale_fill_manual(labels=c("Original 40", "Rattus 8", "Mus 14", "New 70", "Aussie 46", "New 27"),
                     values=c("#920000", "#333333", "#006ddb", "#db6d00", "#b66dff", "#db6d00")) +
  #ggtitle("Spades assembly counts") +
  labs(x="Sequencing batch", y="Avg. depth to assembly") +
  scale_x_discrete(labels=c("Original 40", "Rattus 8", "Mus 14", "New 70", "Aussie 46", "New 27")) +
  bartheme() +
  theme(legend.position="none",
        axis.text.x = element_text(angle=25, hjust=1))

print(depth_p)
```

### Comparison of read depth mapped to assembly vs. to mouse

```{r fig5, out.width="100%", fig.align="center", warning=FALSE}
####################
# Read depth

asm_depth = subset(spades_data, select=c(Label, avg.depth.per.base..samtools.))
names(asm_depth)[2] = "Depth"
asm_depth$Mapping = "Assembly"
asm_avg = signif(mean(asm_depth$Depth, na.rm=T), 3)

mouse_depth = subset(spades_data, select=c(Label, avg.depth.per.base.to.mouse.targets..samtools.))
names(mouse_depth)[2] = "Depth"
mouse_depth$Mapping = "Mouse targets"
mouse_avg = signif(mean(mouse_depth$Depth, na.rm=T), 3)

paired_depth = rbind(asm_depth, mouse_depth)

# depth_p = ggpaired(paired_depth, x="Mapping", y="Depth",
#                  fill="Mapping", palette=corecol(pal="trek", numcol=2), color="#333333",
#                  line.color="#b3b3b3", line.size = 0.4,
#                  width=0.4, point.size=2,
#                  xlab="Mapping", ylab="Average read depth") +
#   stat_compare_means(paired=TRUE, label.x=1) +
#   bartheme() +
#   theme(legend.position="none")
# 
# print(depth_p)

cols = corecol(pal="trek", numcol=2, offset=1)

depth_p = ggplot(paired_depth, aes(x=Mapping, y=Depth, fill=Mapping)) +
  geom_quasirandom(size=2, alpha=0.7, width=0.25, color="#d3d3d3") +
  geom_boxplot(outlier.shape=NA, alpha=0.7, width=0.5) +
  xlab("Mapping") +
  ylab("Average read depth") +
  scale_fill_manual(limits=c("Assembly", "Mouse targets"), values=cols) +
  geom_hline(yintercept=asm_avg, linetype=2, size=1, color=cols[1]) +
  annotate("text", x=2.4, y=asm_avg-3, label=asm_avg, color=cols[1], size=4) +
  geom_hline(yintercept=mouse_avg, linetype=2, size=1, color=cols[2]) +
  annotate("text", x=2.4, y=mouse_avg+4, label=mouse_avg, color=cols[2], size=4) +
  bartheme() +
  theme(legend.position="none")
print(depth_p)

```

### Comparison of # of reads mapped to assembly vs. to mouse

```{r fig6, out.width="100%", fig.align="center", warning=FALSE}
####################
# Read depth

asm_reads = subset(spades_data, select=c(Label, number.reads.mapped))
names(asm_reads)[2] = "Num.reads"
asm_reads$Mapping = "Assembly"
asm_avg_reads = signif(mean(asm_reads$Num.reads, na.rm=T), 3)

mouse_reads = subset(spades_data, select=c(Label, number.reads.mapped.mouse.targets))
names(mouse_reads)[2] = "Num.reads"
mouse_reads$Mapping = "Mouse targets"
mouse_avg_reads = signif(mean(mouse_reads$Num.reads, na.rm=T), 3)

paired_reads = rbind(asm_reads, mouse_reads)

# depth_p = ggpaired(paired_depth, x="Mapping", y="Depth",
#                  fill="Mapping", palette=corecol(pal="trek", numcol=2), color="#333333",
#                  line.color="#b3b3b3", line.size = 0.4,
#                  width=0.4, point.size=2,
#                  xlab="Mapping", ylab="Average read depth") +
#   stat_compare_means(paired=TRUE, label.x=1) +
#   bartheme() +
#   theme(legend.position="none")
# 
# print(depth_p)

cols = corecol(pal="trek", numcol=2, offset=1)

reads_p = ggplot(paired_reads, aes(x=Mapping, y=Num.reads, fill=Mapping)) +
  geom_quasirandom(size=2, alpha=0.7, width=0.25, color="#d3d3d3") +
  geom_boxplot(outlier.shape=NA, alpha=0.7, width=0.5) +
  xlab("Mapping") +
  ylab("# reads mapped") +
  scale_fill_manual(limits=c("Assembly", "Mouse targets"), values=cols) +
  geom_hline(yintercept=asm_avg_reads, linetype=2, size=1, color=cols[1]) +
  annotate("text", x=2.4, y=asm_avg_reads-2000000, label=asm_avg_reads, color=cols[1], size=4) +
  geom_hline(yintercept=mouse_avg_reads, linetype=2, size=1, color=cols[2]) +
  annotate("text", x=2.4, y=mouse_avg_reads+3000000, label=mouse_avg_reads, color=cols[2], size=4) +
  bartheme() +
  theme(legend.position="none")
print(reads_p)

```

### Comparison of % of reads mapped to assembly vs. to mouse

```{r fig7, out.width="100%", fig.align="center", warning=FALSE}
####################
# Read depth

asm_perc_reads = subset(spades_data, select=c(Label, percent.reads.mapped))
names(asm_perc_reads)[2] = "Perc.reads"
asm_perc_reads$Mapping = "Assembly"
asm_avg_perc = signif(mean(asm_perc_reads$Perc.reads, na.rm=T), 3)

mouse_perc_reads = subset(spades_data, select=c(Label, percent.reads.mapped.mouse.targets))
names(mouse_perc_reads)[2] = "Perc.reads"
mouse_perc_reads$Mapping = "Mouse targets"
mouse_avg_perc = signif(mean(mouse_perc_reads$Perc.reads, na.rm=T), 3)

paired_perc = rbind(asm_perc_reads, mouse_perc_reads)

# depth_p = ggpaired(paired_depth, x="Mapping", y="Depth",
#                  fill="Mapping", palette=corecol(pal="trek", numcol=2), color="#333333",
#                  line.color="#b3b3b3", line.size = 0.4,
#                  width=0.4, point.size=2,
#                  xlab="Mapping", ylab="Average read depth") +
#   stat_compare_means(paired=TRUE, label.x=1) +
#   bartheme() +
#   theme(legend.position="none")
# 
# print(depth_p)

cols = corecol(pal="trek", numcol=2, offset=1)

perc_p = ggplot(paired_perc, aes(x=Mapping, y=Perc.reads, fill=Mapping)) +
  geom_quasirandom(size=2, alpha=0.7, width=0.25, color="#d3d3d3") +
  geom_boxplot(outlier.shape=NA, alpha=0.7, width=0.5) +
  xlab("Mapping") +
  ylab("% reads mapped") +
  scale_fill_manual(limits=c("Assembly", "Mouse targets"), values=cols) +
  geom_hline(yintercept=asm_avg_perc, linetype=2, size=1, color=cols[1]) +
  annotate("text", x=2.4, y=asm_avg_perc-1, label=asm_avg_perc, color=cols[1], size=4) +
  geom_hline(yintercept=mouse_avg_perc, linetype=2, size=1, color=cols[2]) +
  annotate("text", x=2.4, y=mouse_avg_perc+1, label=mouse_avg_perc, color=cols[2], size=4) +
  bartheme() +
  theme(legend.position="none")
print(perc_p)

```

```{r footer}
cat("Page last updated:", format(Sys.time(), "%m/%d/%Y %H:%M:%S %Z"))
htmltools::includeHTML("../html-chunks/rmd_footer.html")
```